### PR TITLE
LoGraAttributor

### PR DESCRIPTION
**Pull request by TheaperDeng**
## Description

This PR introduce LoGraAttributor to `dattri`

### 1. Motivation and Context
We want to include LoGraAttributor as a stable and efficient choice for post-training TDA methods

### 2. Summary of the change
Users API
```python
def f(model, batch, device):
    inputs, targets = batch
    inputs = inputs.to(device)
    targets = targets.to(device)
    outputs = model(inputs)
    return nn.functional.cross_entropy(outputs, targets)

task = AttributionTask(
    model=model,
    loss_func=f,
    checkpoints=["checkpoint_0"],
)

attributor = LoGraAttributor(
    task=task,
    device=args.device
)

attributor.cache(dataloader_train)
score = attributor.attribute(dataloader_train, dataloader_test)
```

### 3. What tests have been added/updated for the change?
- [ ] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.
- [ ] Application test: If you wrote an example for the toolkit, this test should be added.
- [ ] Document test: If you added an external API, then you should check if the document is correctly generated.

### CODE REVIEW

**Comment by jiaqima** [comment]
Please address the failed tests.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,19 +0,0 @@
...
-  -   id: check-added-large-files
-- repo: https://github.com/astral-sh/ruff-pre-commit
-  # Ruff version.
-  rev: v0.4.1
-  hooks:
-    # Run the linter.
-    - id: ruff
-      args: [ --fix ]
-    # Run the formatter.
-    - id: ruff-format
```
Comment (TheaperDeng) [inline .pre-commit-config.yaml:19]: The file should be added back.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -23,19 +23,6 @@ Install the necessary dependencies:
...
-```bash
-conda create -n dattri python=3.10
-conda activate dattri
-
-conda install -c "nvidia/label/cuda-11.8.0" cuda-toolkit
-pip3 install torch==2.1.0 --index-url https://download.pytorch.org/whl/cu118
-
-pip install -e .
-```
-
```
Comment (TheaperDeng) [inline CONTRIBUTING.md:38]: These lines should be added back.

File: assets/images/logo.png
Comment (TheaperDeng) [inline assets/images/logo.png:1]: This file should use the original logo.

File: batch_metadata.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{"batch_info":{"0":{"start_idx":0,"sample_count":4},"1":{"start_idx":4,"sample_count":4},"2":{"start_idx":8,"sample_count":2}},"layer_names":["linear"],"layer_dims":[8],"total_proj_dim":8,"total_samples":10,"timestamp":1757897081.690028}
```
Comment (jiaqima) [inline batch_metadata.json:1]: What's the purpose of adding this file?

File: dattri/algorithm/tracin.py [hunk 4]
```diff
@@ -19,14 +19,6 @@
 from .base import BaseAttributor
 from .utils import _check_shuffle
 
-DEFAULT_PROJECTOR_KWARGS = {
```
Comment (jiaqima) [inline dattri/algorithm/tracin.py:22]: Why do we need to change tracin (and trak) for adding logra?

File: batch_metadata.json (hunk 3 above)
Comment (haochend413) [reply batch_metadata.json:1]: This is auto-generated for bookkeeping. I'll have it removed in future commits.

File: dattri/algorithm/tracin.py (hunk 4 above)
Comment (jiaqima) [reply dattri/algorithm/tracin.py:22]: Thanks for the clarification. When you get everything ready in the end, maybe it's worth opening a clean new PR to have a clearer change log.

**Comment by TheaperDeng** [comment]
@haochend413 Let's rename `logra_utils/` as `logra` and move `logra.py` to the folder. At the same time, we can add `LoGraAttributor` to https://github.com/sleepymalc/dattri/blob/2d759a125189d8e1051f3a6fe5d3a8e8ff42589c/dattri/algorithm/__init__.py#L3-L17

**Comment by haochend413** [comment]
> @haochend413 Let's rename `logra_utils/` as `logra` and move `logra.py` to the folder. At the same time, we can add `LoGraAttributor` to https://github.com/sleepymalc/dattri/blob/2d759a125189d8e1051f3a6fe5d3a8e8ff42589c/dattri/algorithm/__init__.py#L3-L17

Do we need to include algorithms other than IF and LoGra? TRAK and TRACIN are not included in this file as well in the main branch.

**Comment by TheaperDeng** [comment]
yes lets add others in as well

File: dattri/algorithm/logra/core/hook.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        # Handle bias term by augmenting input with ones
+        if layer.bias is not None:
+            ones = torch.ones(
+                input_features_flat.size(0),
+                1,
+                device=input_features_flat.device,
+                dtype=input_features_flat.dtype,
+            )
+            input_features_flat = torch.cat([input_features_flat, ones], dim=1)
+
```
Comment (sleepymalc) [inline dattri/algorithm/logra/core/hook.py:278]: We should project each component here using the two projectors of this layer, respectively.

**Review by sleepymalc — COMMENTED** [review]
Generally, I think the gradient compression logic in `hook.py` is incorrect: it directly reconstructs the gradient (using outer product) from the components and then projects. Instead, we should initialize two projectors for each layer, and then project the individual components first, then reconstruct the gradient using the outer product.

File: dattri/algorithm/logra/core/hook.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+            per_sample: Whether to compute per-sample gradients
+
+        Returns:
+            Projected gradient tensor
+        """
+        input_features = self.inputs[idx]
+        is_3d = input_features.dim() == 3
+
+        # Get projector for this layer
+        projector = (
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply dattri/algorithm/logra/core/hook.py:245]: The implementation directly follow your branch. I remember previously we have 2 projectors (like 4-5 months ago), could this be a false rollback？
